File: c++/src/netstream-sender.cpp
/**
 *
 *
 * Copyright (c) 2010-2011 University of Luxembourg
 *
 *
 * @file netstream_sender.cc
 * @date 2011-08-22
 *
 * @author Yoann Pigné
 */


#include "netstream-sender.h"

namespace netstream{
  
NetStreamSender::NetStreamSender(const GS_STRING & host, GS_INT port):
_stream_name("default"),_host(host),_port(port),_stream(),_socket(host,port),debug(false)  {
  init();
}

NetStreamSender::NetStreamSender(GS_INT port):
_stream_name("default"),_host("localhost"),_port(port),_stream(),_socket("localhost",port),debug(false)  {
  init();
}

NetStreamSender::NetStreamSender(const GS_STRING & stream, const GS_STRING & host, GS_INT port):
_stream_name(stream),_host(host),_port(port),_stream(),_socket(host,port),debug(false)
{
  init();
}
NetStreamSender::NetStreamSender(const GS_STRING & stream, const GS_STRING & host, GS_INT port, GS_BOOL debug):
_stream_name(stream),_host(host),_port(port),_stream(),_socket(host,port),debug(debug)
{
  init();
}


void NetStreamSender::init() 
{ 
  _stream.writeString(_stream_name);
  
  int wait_for_server = 1;
  while(wait_for_server){
    try{
      _socket.connect();
      if(wait_for_server > 1)
            std::cout<<std::endl<<"Connection established."<<std::endl;
      wait_for_server = 0;
    } catch(NetStreamSocketException e){
      if(wait_for_server == 1) 
        std::cout<<"No available connection on "<<_host<<":"<<_port<<". Waiting.";
      else
        std::cout<<"."<<std::flush;
      wait_for_server++;
      sleep(1);
    }
  }
}

// ===========================================
// = Values type guess (templates) =
// ===========================================
GS_INT NetStreamSender::_getType(GS_CHAR object){
  if(debug){
    cerr<<"NetStreamSender: _getType : char"<<endl;
  }
  return TYPE_BYTE;
}

GS_INT NetStreamSender::_getType(GS_BOOL object){
if(debug){
  cerr<<"NetStreamSender: _getType : bool"<<endl;
}
return TYPE_BOOLEAN;}

GS_INT NetStreamSender::_getType(GS_INT object){
if(debug){
  cerr<<"NetStreamSender: _getType : int"<<endl;
}
return TYPE_INT;}

GS_INT NetStreamSender::_getType(GS_LONG object){
if(debug){
  cerr<<"NetStreamSender: _getType : long"<<endl;
}
return TYPE_LONG;}

GS_INT NetStreamSender::_getType(GS_FLOAT object){
if(debug){
  cerr<<"NetStreamSender: _getType : float"<<endl;
}
return TYPE_FLOAT;}

GS_INT NetStreamSender::_getType(GS_DOUBLE object){
if(debug){
  cerr<<"NetStreamSender: _getType : double"<<endl;
}
return TYPE_DOUBLE;}

GS_INT NetStreamSender::_getType(const GS_STRING & object){
if(debug){
  cerr<<"NetStreamSender: _getType : string"<<endl;
}
return TYPE_STRING;}

GS_INT NetStreamSender::_getType(const vector<GS_CHAR> & object){
if(debug){
  cerr<<"NetStreamSender: _getType : char* "<<endl;
}
return TYPE_BYTE_ARRAY;}

GS_INT NetStreamSender::_getType(const vector <GS_BOOL> & object){
if(debug){
  cerr<<"NetStreamSender: _getType : bool*"<<endl;
}
return TYPE_BOOLEAN_ARRAY;}

GS_INT NetStreamSender::_getType(const vector<GS_INT> & object){
if(debug){
  cerr<<"NetStreamSender: _getType : int*"<<endl;
}
return TYPE_INT_ARRAY;}

GS_INT NetStreamSender::_getType(const vector<GS_LONG> & object){
if(debug){
  cerr<<"NetStreamSender: _getType : long*"<<endl;
}
return TYPE_LONG_ARRAY;}

GS_INT NetStreamSender::_getType(const vector<GS_FLOAT> & object){
if(debug){
  cerr<<"NetStreamSender: _getType : float*"<<endl;
}
return TYPE_FLOAT_ARRAY;}

GS_INT NetStreamSender::_getType(const vector <GS_DOUBLE> & object){
if(debug){
  cerr<<"NetStreamSender: _getType : double*"<<endl;
}
return TYPE_DOUBLE_ARRAY;}


// =================
// = data encoding =
// =================
void NetStreamSender::_encode(NetStreamStorage & event, GS_CHAR value){
  event.writeByte((GS_INT)value);
}

void NetStreamSender::_encode(NetStreamStorage & event, GS_BOOL value){
  event.writeByte(value?1:0);
}

void NetStreamSender::_encode(NetStreamStorage & event, GS_INT value){
  event.writeInt(value);
}

void NetStreamSender::_encode(NetStreamStorage & event, GS_LONG value){
  event.writeLong(value);
}

void NetStreamSender::_encode(NetStreamStorage & event, GS_FLOAT value){
  event.writeFloat(value);
}

void NetStreamSender::_encode(NetStreamStorage & event, GS_DOUBLE value){
  event.writeDouble(value);
}

void NetStreamSender::_encode(NetStreamStorage & event, const GS_STRING & value){
  event.writeString(value);
}

void NetStreamSender::_encode(NetStreamStorage & event, const vector<GS_CHAR> & value){
  event.writeInt(value.size());
  for(vector<char>::const_iterator i = value.begin(); i != value.end(); i++){
    event.writeByte((*i));
  }
}

void NetStreamSender::_encode(NetStreamStorage & event, const vector<GS_BOOL> & value){
  event.writeInt(value.size());
  for(vector<bool>::const_iterator i = value.begin(); i != value.end(); i++){
    event.writeByte((*i));
  }
}

void NetStreamSender::_encode(NetStreamStorage & event, const vector<GS_INT> & value){
  event.writeInt(value.size());
  for(vector<int>::const_iterator i = value.begin(); i != value.end(); i++){
    event.writeInt((*i));
  }
}

void NetStreamSender::_encode(NetStreamStorage & event, const vector<GS_LONG> & value){
  event.writeInt(value.size());
  for(vector<GS_LONG>::const_iterator i = value.begin(); i != value.end(); i++){
    event.writeLong((*i));
  }  
}

void NetStreamSender::_encode(NetStreamStorage & event, const vector<GS_FLOAT> & value){
  event.writeInt(value.size());
  for(vector<GS_FLOAT>::const_iterator i = value.begin(); i != value.end(); i++){
    event.writeFloat((*i));
  }
}

void NetStreamSender::_encode(NetStreamStorage & event, const vector<GS_DOUBLE> &  value){
  event.writeInt(value.size());
  for(vector<GS_DOUBLE>::const_iterator i = value.begin(); i != value.end(); i++){
    event.writeDouble((*i));
  }
}


void NetStreamSender::_sendEvent(NetStreamStorage & event){
  
  if(debug){
    cout<<event<<endl;
  }
  try{
    _socket.sendExact(_stream+event);
  }
  catch (NetStreamSocketException &e)
  {
    cout << "Error while sending message: " << e.what();
  }
}

// ==================
// = Element events =
// ================== 
void NetStreamSender::addNode(const GS_STRING & source_id, GS_LONG time_id, const GS_STRING & node_id){
  NetStreamStorage event = NetStreamStorage();
  event.writeByte(EVENT_ADD_NODE);
  event.writeString(source_id);
  event.writeUnsignedVarInt(time_id);
  event.writeString(node_id);
  _sendEvent(event);
}

void NetStreamSender::removeNode(const GS_STRING & source_id, GS_LONG time_id, const GS_STRING & node_id){
  NetStreamStorage event = NetStreamStorage();
  event.writeByte(EVENT_DEL_NODE);
  event.writeString(source_id);
  event.writeUnsignedVarInt(time_id);
  event.writeString(node_id);
  _sendEvent(event);
}

void NetStreamSender::addEdge(const GS_STRING & source_id, GS_LONG time_id, const GS_STRING & edge_id, const GS_STRING & from_node, const GS_STRING & to_node, GS_BOOL directed){
  NetStreamStorage event = NetStreamStorage();
  event.writeByte(EVENT_ADD_EDGE);
  event.writeString(source_id);
  event.writeUnsignedVarInt(time_id);
  event.writeString(edge_id);
  event.writeString(from_node);
  event.writeString(to_node);
  event.writeByte(directed?1:0);
  _sendEvent(event);
}

void NetStreamSender::removeEdge(const GS_STRING & source_id, GS_LONG time_id, const GS_STRING & edge_id){
  NetStreamStorage event = NetStreamStorage();
  event.writeByte(EVENT_DEL_EDGE);
  event.writeString(source_id);
  event.writeUnsignedVarInt(time_id);
  event.writeString(edge_id);
  _sendEvent(event);
}

void NetStreamSender::stepBegins(const GS_STRING & source_id, GS_LONG time_id, GS_DOUBLE timestamp){
  NetStreamStorage event = NetStreamStorage();
  event.writeByte(EVENT_STEP);
  event.writeString(source_id);
  event.writeUnsignedVarInt(time_id);
  event.writeDouble(timestamp);
  _sendEvent(event);
}

void NetStreamSender::graphClear(const GS_STRING & source_id, GS_LONG time_id){
  NetStreamStorage event = NetStreamStorage();
  event.writeByte(EVENT_CLEARED);
  event.writeString(source_id);
  event.writeUnsignedVarInt(time_id);
  _sendEvent(event);  
}

// =====================
// = Attributes events =
// ===================== 

void NetStreamSender::removeNodeAttribute(const GS_STRING & source_id, GS_LONG time_id, const GS_STRING & node_id, const GS_STRING & attribute){
  NetStreamStorage event = NetStreamStorage();
  event.writeByte(EVENT_DEL_NODE_ATTR);
  event.writeString(source_id);
  event.writeUnsignedVarInt(time_id);
  event.writeString(node_id);
  event.writeString(attribute);
  _sendEvent(event);
}

void NetStreamSender::removeGraphAttribute(const GS_STRING & source_id, GS_LONG time_id, const GS_STRING & attribute){
  NetStreamStorage event = NetStreamStorage();
  event.writeByte(EVENT_DEL_GRAPH_ATTR);
  event.writeString(source_id);
  event.writeUnsignedVarInt(time_id);
  event.writeString(attribute);
  _sendEvent(event);
}

void NetStreamSender::removeEdgeAttribute(const GS_STRING & source_id, GS_LONG time_id, const GS_STRING & edge_id, const GS_STRING & attribute){
  NetStreamStorage event = NetStreamStorage();
  event.writeByte(EVENT_DEL_EDGE_ATTR);
  event.writeString(source_id);
  event.writeUnsignedVarInt(time_id);
  event.writeString(edge_id);
  event.writeString(attribute);
  _sendEvent(event);
}

}

File: csharp/src/NetStreamSender.cs
/*
 * Copyright 2006 - 2013
 *     Stefan Balev     <stefan.balev@graphstream-project.org>
 *     Julien Baudry    <julien.baudry@graphstream-project.org>
 *     Antoine Dutot    <antoine.dutot@graphstream-project.org>
 *     Yoann Pigné      <yoann.pigne@graphstream-project.org>
 *     Guilhelm Savin   <guilhelm.savin@graphstream-project.org>
 * 
 * This file is part of GraphStream <http://graphstream-project.org>.
 * 
 * GraphStream is a library whose purpose is to handle static or dynamic
 * graph, create them from scratch, file or any source and display them.
 * 
 * This program is free software distributed under the terms of two licenses, the
 * CeCILL-C license that fits European law, and the GNU Lesser General Public
 * License. You can  use, modify and/ or redistribute the software under the terms
 * of the CeCILL-C license as circulated by CEA, CNRS and INRIA at the following
 * URL <http://www.cecill.info> or under the terms of the GNU LGPL as published by
 * the Free Software Foundation, either version 3 of the License, or (at your
 * option) any later version.
 * 
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
 * PARTICULAR PURPOSE.  See the GNU Lesser General Public License for more details.
 * 
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 * 
 * The fact that you are presently reading this means that you have had
 * knowledge of the CeCILL-C and LGPL licenses and that you accept their terms.
 */
 
/**
 * Copyright (c) 2015 Technische Universtität München, Germany
 *
 * NetStreamSender.cs
 * @since Feb 25, 2015
 *
 * @author Benjamin Krämer <benjamin.kraemer@in.tum.de>
 *
 */
using System;
using System.IO;
using System.Net;
using System.Net.Sockets;
using System.Text;

namespace Netstream
{
    internal class NetStreamSender
    {
        private readonly byte[] _streamIdArray;
        private readonly String _host;
        private readonly int _port;
        private Socket _socket;
        private BufferedStream _outStream;
        private NetStreamPacker _packer = new DefaultPacker();

        private class DefaultPacker : NetStreamPacker
        {
            private readonly NetStreamStorage _sizeBuffer = new NetStreamStorage(sizeof (int));

            public override NetStreamStorage PackMessage(NetStreamStorage buffer, int startIndex, int endIndex)
            {
                return buffer;
            }

            public override NetStreamStorage PackMessageSize(int capacity)
            {
                _sizeBuffer.Position = 0;
                byte[] bytes = BitConverter.GetBytes(capacity);
                if (BitConverter.IsLittleEndian)
                    Array.Reverse(bytes);
                _sizeBuffer.Write(bytes, 0, bytes.Length);
                return _sizeBuffer;
            }

        };

        public NetStreamSender(int port)
            : this("default", "localhost", port)
        {

        }

        public NetStreamSender(String host, int port)
            : this("default", host, port)
        {
        }

        public NetStreamSender(String stream, String host, int port)
        {
            _host = host;
            _port = port;
            _streamIdArray = Encoding.UTF8.GetBytes(stream);

            Connect();
        }

        /**
         * Sets an optional NetStreamPaker whose "pack" method will be called on
         * each message.
         * 
         * a Packer can do extra encoding on the all byte array message, it may also
         * crypt things.
         * 
         * @param paker
         *            The packer object
         */

        public void SetPacker(NetStreamPacker paker)
        {
            _packer = paker;
        }

        public void RemovePacker()
        {
            _packer = new DefaultPacker();
        }

        protected void Connect()
        {
            _socket = new Socket(SocketType.Stream, ProtocolType.Tcp);
            _socket.Connect(new DnsEndPoint(_host, _port));
            _outStream = new BufferedStream(new NetworkStream(_socket));
        }

        /**
         * @param buff
         */

        private void DoSend(NetStreamStorage buff)
        {
            if (!_socket.Connected)
            {
                Console.Error.WriteLine("NetStreamSender : can't send. The socket is closed.");
            }
            else
            {
                buff.Flip();
                NetStreamStorage buffer = _packer.PackMessage(buff);
                NetStreamStorage sizeBuffer = _packer.PackMessageSize(buffer.Capacity);
                buff.Position = 0;

                // real sending
                try
                {
                    _outStream.Write(sizeBuffer.ToArray(), 0, sizeBuffer.Capacity);
                    _outStream.Write(buffer.ToArray(), 0, buffer.Capacity);
                    _outStream.Flush();
                }
                catch (IOException e)
                {
                    Console.Error.WriteLine(e.StackTrace);
                }
            }
        }

        protected void AddAttribute(String sourceId, ulong timeId, String attribute, Object value, NetStreamEvent e)
        {
            NetStreamStorage buff = new NetStreamStorage().
                EncodeArray(_streamIdArray).
                EncodeEvent(e).
                EncodeString(sourceId).
                EncodeNative(timeId).
                EncodeString(attribute).
                EncodeValueWithType(value);
            DoSend(buff);
        }

        protected void ChangeAttribute(String sourceId, ulong timeId, String attribute, Object oldValue, Object newValue,
            NetStreamEvent e)
        {
            NetStreamStorage buff = new NetStreamStorage().
                EncodeArray(_streamIdArray).
                EncodeEvent(e).
                EncodeString(sourceId).
                EncodeNative(timeId).
                EncodeString(attribute).
                EncodeValueWithType(oldValue).
                EncodeValueWithType(newValue);
            DoSend(buff);
        }

        protected void RemoveAttribute(String sourceId, ulong timeId, String attribute, NetStreamEvent e)
        {
            NetStreamStorage buff = new NetStreamStorage().
                EncodeArray(_streamIdArray).
                EncodeEvent(e).
                EncodeString(sourceId).
                EncodeNative(timeId).
                EncodeString(attribute);
            DoSend(buff);
        }

        /*
         * (non-Javadoc)
         * 
         * @see
         * org.graphstream.stream.AttributeSink#graphAttributeAdded(java.lang.String
         * , long, java.lang.String, java.lang.Object)
         */
        public void AddGraphAttribute(String sourceId, ulong timeId, String attribute, Object value)
        {
            AddAttribute(sourceId, timeId, attribute, value, NetStreamEvent.AddGraphAttr);
        }

        /*
	     * (non-Javadoc)
	     * 
	     * @see
	     * org.graphstream.stream.AttributeSink#graphAttributeChanged(java.lang.
	     * String, long, java.lang.String, java.lang.Object, java.lang.Object)
	     */
        public void ChangeGraphAttribute(String sourceId, ulong timeId, String attribute, Object oldValue,
            Object newValue)
        {
            ChangeAttribute(sourceId, timeId, attribute, oldValue, newValue, NetStreamEvent.ChgGraphAttr);
        }

        /*
         * (non-Javadoc)
         * 
         * @see
         * org.graphstream.stream.AttributeSink#graphAttributeRemoved(java.lang.
         * String, long, java.lang.String)
         */
        public void RemoveGraphAttribute(String sourceId, ulong timeId, String attribute)
        {
            RemoveAttribute(sourceId, timeId, attribute, NetStreamEvent.DelGraphAttr);
        }

        /*
         * (non-Javadoc)
         * 
         * @see
         * org.graphstream.stream.AttributeSink#nodeAttributeAdded(java.lang.String,
         * long, java.lang.String, java.lang.String, java.lang.Object)
         */
        public void AddNodeAttribute(String sourceId, ulong timeId, String nodeId,
            String attribute, Object value)
        {
            AddAttribute(sourceId, timeId, attribute, value, NetStreamEvent.AddNodeAttr);
        }

        /*
         * (non-Javadoc)
         * 
         * @see
         * org.graphstream.stream.AttributeSink#nodeAttributeChanged(java.lang.String
         * , long, java.lang.String, java.lang.String, java.lang.Object,
         * java.lang.Object)
         */
        public void ChangeNodeAttribute(String sourceId, ulong timeId,
            String nodeId, String attribute, Object oldValue, Object newValue)
        {
            ChangeAttribute(sourceId, timeId, attribute, oldValue, newValue, NetStreamEvent.ChgNodeAttr);
        }

        /*
         * (non-Javadoc)
         * 
         * @see
         * org.graphstream.stream.AttributeSink#nodeAttributeRemoved(java.lang.String
         * , long, java.lang.String, java.lang.String)
         */
        public void RemoveNodeAttribute(String sourceId, ulong timeId,
            String nodeId, String attribute)
        {
            RemoveAttribute(sourceId, timeId, attribute, NetStreamEvent.DelNodeAttr);
        }

        /*
         * (non-Javadoc)
         * 
         * @see
         * org.graphstream.stream.AttributeSink#edgeAttributeAdded(java.lang.String,
         * long, java.lang.String, java.lang.String, java.lang.Object)
         */
        public void AddEdgeAttribute(String sourceId, ulong timeId, String edgeId,
            String attribute, Object value)
        {
            AddAttribute(sourceId, timeId, attribute, value, NetStreamEvent.AddEdgeAttr);
        }

        /*
         * (non-Javadoc)
         * 
         * @see
         * org.graphstream.stream.AttributeSink#edgeAttributeChanged(java.lang.String
         * , long, java.lang.String, java.lang.String, java.lang.Object,
         * java.lang.Object)
         */
        public void ChangeEdgeAttribute(String sourceId, ulong timeId,
            String edgeId, String attribute, Object oldValue, Object newValue)
        {
            ChangeAttribute(sourceId, timeId, attribute, oldValue, newValue, NetStreamEvent.ChgEdgeAttr);
        }

        /*
         * (non-Javadoc)
         * 
         * @see
         * org.graphstream.stream.AttributeSink#edgeAttributeRemoved(java.lang.String
         * , long, java.lang.String, java.lang.String)
         */
        public void RemoveEdgeAttribute(String sourceId, ulong timeId,
            String edgeId, String attribute)
        {
            RemoveAttribute(sourceId, timeId, attribute, NetStreamEvent.DelEdgeAttr);
        }

        /*
         * (non-Javadoc)
         * 
         * @see org.graphstream.stream.ElementSink#nodeAdded(java.lang.String, long,
         * java.lang.String)
         */
        public void AddNode(String sourceId, ulong timeId, String nodeId)
        {
            NetStreamStorage buff = new NetStreamStorage().
                EncodeArray(_streamIdArray).
                EncodeEvent(NetStreamEvent.AddNode).
                EncodeString(sourceId).
                EncodeNative(timeId).
                EncodeString(nodeId);
            DoSend(buff);
        }

        /*
         * (non-Javadoc)
         * 
         * @see org.graphstream.stream.ElementSink#nodeRemoved(java.lang.String,
         * long, java.lang.String)
         */
        public void RemoveNode(String sourceId, ulong timeId, String nodeId)
        {
            NetStreamStorage buff = new NetStreamStorage().
                EncodeArray(_streamIdArray).
                EncodeEvent(NetStreamEvent.DelNode).
                EncodeString(sourceId).
                EncodeNative(timeId).
                EncodeString(nodeId);
            DoSend(buff);
        }

        /*
         * (non-Javadoc)
         * 
         * @see org.graphstream.stream.ElementSink#edgeAdded(java.lang.String, long,
         * java.lang.String, java.lang.String, java.lang.String, boolean)
         */
        public void AddEdge(String sourceId, ulong timeId, String edgeId,
            String fromNodeId, String toNodeId, bool directed)
        {
            NetStreamStorage buff = new NetStreamStorage().
                EncodeArray(_streamIdArray).
                EncodeEvent(NetStreamEvent.AddEdge).
                EncodeString(sourceId).
                EncodeNative(timeId).
                EncodeString(edgeId).
                EncodeString(fromNodeId).
                EncodeString(toNodeId).
                EncodeNative(directed);
            DoSend(buff);
        }

        /*
         * (non-Javadoc)
         * 
         * @see org.graphstream.stream.ElementSink#edgeRemoved(java.lang.String,
         * long, java.lang.String)
         */
        public void RemoveEdge(String sourceId, ulong timeId, String edgeId)
        {
            NetStreamStorage buff = new NetStreamStorage().
                EncodeArray(_streamIdArray).
                EncodeEvent(NetStreamEvent.DelEdge).
                EncodeString(sourceId).
                EncodeNative(timeId).
                EncodeString(edgeId);
            DoSend(buff);
        }

        /*
         * (non-Javadoc)
         * 
         * @see org.graphstream.stream.ElementSink#graphCleared(java.lang.String,
         * long)
         */
        public void GraphClear(String sourceId, ulong timeId)
        {
            NetStreamStorage buff = new NetStreamStorage().
                EncodeArray(_streamIdArray).
                EncodeEvent(NetStreamEvent.Cleared).
                EncodeString(sourceId).
                EncodeNative(timeId);
            DoSend(buff);
        }

        /*
         * (non-Javadoc)
         * 
         * @see org.graphstream.stream.ElementSink#stepBegins(java.lang.String,
         * long, double)
         */
        public void StepBegins(String sourceId, ulong timeId, double step)
        {
            NetStreamStorage buff = new NetStreamStorage().
                EncodeArray(_streamIdArray).
                EncodeEvent(NetStreamEvent.Step).
                EncodeString(sourceId).
                EncodeNative(timeId).
                EncodeNative(step);
            DoSend(buff);
        }

        /**
         * Force the connection to close (properly) with the server
         * 
         * @throws IOException
         */
        public void Close()
        {
            _socket.Close();
        }
    }
}
